### PR TITLE
Tessil/robin-map implemented instead unordered_map and unordered_set

### PR DESCRIPTION
**Pull request by boludoz**
Robin map and robin set implemented, like Skyline, for more performance.

### CODE REVIEW

**Comment by AliciaBurrito** [comment]
@boludoz Do you have any examples of the improved performance? eg. games to test, or what numbers you got?

**Comment by FearlessTobi** [comment]
Blindly replacing every unordered_map with robin_map, without providing any performance numbers will not get accepted here.

**Comment by boludoz** [comment]
> Blindly replacing every unordered_map with robin_map, without providing any performance numbers will not get accepted here.

image

**Comment by boludoz** [comment]
```
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include "tsl/robin_map.h"
#include "tsl/robin_set.h"
#include <chrono>

int main() {
    std::unordered_map<int, int> unorderedMap;
    tsl::robin_map<int, int> robinMap;
    std::unordered_set<int> unorderedSet;
    tsl::robin_set<int> robinSet;

    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000000; ++i) {
        unorderedMap[i] = i * 2;
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed_seconds = end - start;

    std::cout << "Time to insert into std::unordered_map: " << elapsed_seconds.count() << " seconds" << std::endl;
    
    start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000000; ++i) {
        robinMap[i] = i * 2;
    }

    end = std::chrono::high_resolution_clock::now();
    elapsed_seconds = end - start;

    std::cout << "Time to insert into tsl::robin_map: " << elapsed_seconds.count() << " seconds" << std::endl;

    start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000000; ++i) {
        unorderedSet.insert(i);
    }

    end = std::chrono::high_resolution_clock::now();
    elapsed_seconds = end - start;

    std::cout << "Time to insert into std::unordered_set: " << elapsed_seconds.count() << " seconds" << std::endl;

    start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000000; ++i) {
        robinSet.insert(i);
    }

    end = std::chrono::high_resolution_clock::now();
    elapsed_seconds = end - start;
    std::cout << "Time to insert into tsl::robin_set: " << elapsed_seconds.count() << " seconds" << std::endl;
    return 0;
}

```

**Comment by boludoz** [comment]
> Blindly replacing every unordered_map with robin_map, without providing any performance numbers will not get accepted here.

I ask you to please reconsider your opinion, perhaps the performance is not too much but it can help. It is a grain of sand that makes the mountain. 30 % more performance in tsl::robin_set for example.

**Comment by boludoz** [comment]
> @boludoz Do you have any examples of the improved performance? eg. games to test, or what numbers you got?

image
I don't give explanations to xenophobic people.

**Comment by FernandoS27** [comment]
I'm sorry our moderators are morons who cannot respect someone's hard work and put you in the hard position of going through benchmarks and stuffs without much necessity. 

Please feel free to reopen.

**Comment by FernandoS27** [comment]
Oh fix clamg format.

**Comment by gidoly** [comment]
i can check if this pr makes difference in games after ci passed. even if doesn't make big difference it is nice to have

**Comment by boludoz** [comment]
> i can check if this pr makes difference in games after ci passed. even if doesn't make big difference it is nice to have

I think the order is fixed there.

**Comment by gidoly** [comment]
getting compiling error on visual studio
devenv_WUDBhnoVfD
i think it is because missing external file as cmake gives me this error
WindowsTerminal_Oe3nlNFJNb

File: externals/CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,9 @@ set(BUILD_SHARED_LIBS OFF)
 # Skip install rules for all externals
 set_directory_properties(PROPERTIES EXCLUDE_FROM_ALL ON)
 
+# Robin-map
+add_subdirectory(robin-map)
```
Comment (Tachi107) [inline externals/CMakeLists.txt:18]: Could you please add `find_package(tsl-robin-map CONFIG)` to the top-level CMakeLists.txt and change this to the following:

```suggestion
if (NOT TARGET tsl::robin_map)
    add_subdirectory(robin-map)
endif()
```

Also, you don't seem to use the `tsl::robin_map` target anywhere in the CMake files, you should probably add `target_link_libraries(<target-name> PRIVATE tsl::robin_map)` in the relevant place(s).

File: .gitmodules [hunk 2]
```diff
@@ -58,3 +58,6 @@
 [submodule "VulkanMemoryAllocator"]
 	path = externals/VulkanMemoryAllocator
 	url = https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator.git
+[submodule "robin-map"]
+	path = externals/robin-map
+	url = https://github.com/Tessil/robin-map
```
Comment (v1993) [inline .gitmodules:63]:
```suggestion
	url = https://github.com/Tessil/robin-map.git
```

Comment (boludoz) [reply .gitmodules:63]: Thanks sir.

Comment (v1993) [reply .gitmodules:63]: Looks like this was not committed?

**Review by v1993 — CHANGES_REQUESTED** [review]
`.gitmodules` still not fixed

**Comment by boludoz** [comment]
> `.gitmodules` still not fixed

Sorry, pushed now.

**Review by v1993 — COMMENTED** [review]
Benchmarking aside - it's worth noting that robin-map has [some differences](https://github.com/Tessil/robin-map#differences-compared-to-stdunordered_map) in its API compared to vanilla `unordered_map`. While most are covered by compile-time checks, a few differences in iterator invalidation should be verified manually to not affect code in question.

**Review by v1993 — COMMENTED** [review]
Benchmarking aside - it's worth noting that robin-map has [some differences](https://github.com/Tessil/robin-map#differences-compared-to-stdunordered_map) in its API compared to vanilla `unordered_map`. While most are covered by compile-time checks, a few differences in iterator invalidation should be verified manually to not affect code in question.

**Comment by boludoz** [comment]
Fixed submodule

**Comment by boludoz** [comment]
If anyone wants to help, I'm having a hard time adding a simple library.

File: src/yuzu/multiplayer/chat_room.h [hunk 3]
```diff
@@ -64,8 +64,8 @@ public slots:
     bool has_mod_perms = false;
     QStandardItemModel* player_list;
     std::unique_ptr<Ui::ChatRoom> ui;
-    std::unordered_set<std::string> block_list;
-    std::unordered_map<std::string, QPixmap> icon_cache;
+    tsl::robin_set<std::string> block_list;
```
Comment (FearlessTobi) [inline src/yuzu/multiplayer/chat_room.h:67]: this is not performance critical at all.
dito for many other places

**Comment by Kelebek1** [comment]
I don't know if there's a way to make an external available to all projects by default, but I think you need to add robin map to each project's `target_link_libraries` in their cmake file. 
E.g src/common/cmakelists.txt has
```cmake
target_link_libraries(common PUBLIC Boost::context Boost::headers fmt::fmt microprofile Threads::Threads)
target_link_libraries(common PRIVATE lz4::lz4 zstd::zstd LLVM::Demangle)
```
You'll want robin-map in there, in the public one I guess? And repeat for each project that you're having use it.

**Review by FearlessTobi — CHANGES_REQUESTED** [review]
You need to provide actual performance comparsions from games for this to have a chance to get approved.

**Comment by boludoz** [comment]
For my part, there is nothing more to contribute, if anyone wants to solve the problem, welcome...

**Pull request closed without merging**